File: beautylens/modules/expo-face-landmarker/ios/ExpoFaceLandmarkerModule.swift
import ExpoModulesCore
import MediaPipeTasksVision
import UIKit

public class ExpoFaceLandmarkerModule: Module {
  private var faceLandmarker: FaceLandmarker?

  public func definition() -> ModuleDefinition {
    Name("ExpoFaceLandmarker")

    AsyncFunction("detectFromImageAsync") { (imageUri: String) throws -> [String: Any] in
      guard let image = Self.loadImage(imageUri: imageUri) else {
        throw Exception(name: "ImageDecodeError", description: "Could not decode image: \(imageUri)")
      }

      let landmarker = try self.getFaceLandmarker()
      let mpImage = try MPImage(uiImage: image)
      let result = try landmarker.detect(image: mpImage)

      guard let faceLandmarks = result.faceLandmarks.first else {
        return [
          "status": "success",
          "face_detected": false,
          "landmarks": [],
          "num_landmarks": 0,
          "image_dimensions": [
            "width": Int(image.size.width),
            "height": Int(image.size.height)
          ],
          "message": "No face detected in image"
        ]
      }

      let width = Double(image.size.width)
      let height = Double(image.size.height)
      let landmarks = faceLandmarks.map { landmark -> [String: Double] in
        [
          "x": Double(landmark.x) * width,
          "y": Double(landmark.y) * height,
          "z": Double(landmark.z) * width
        ]
      }

      let xs = landmarks.map { $0["x"] ?? 0 }
      let ys = landmarks.map { $0["y"] ?? 0 }
      let minX = xs.min() ?? 0
      let maxX = xs.max() ?? 0
      let minY = ys.min() ?? 0
      let maxY = ys.max() ?? 0

      return [
        "status": "success",
        "face_detected": true,
        "landmarks": landmarks,
        "num_landmarks": landmarks.count,
        "bbox": [
          "x": minX,
          "y": minY,
          "width": maxX - minX,
          "height": maxY - minY
        ],
        "image_dimensions": [
          "width": Int(image.size.width),
          "height": Int(image.size.height)
        ]
      ]
    }
  }

  private func getFaceLandmarker() throws -> FaceLandmarker {
    if let faceLandmarker {
      return faceLandmarker
    }

    guard let modelPath = Bundle.main.path(forResource: "face_landmarker", ofType: "task") else {
      throw Exception(
        name: "MissingModel",
        description: "Missing face_landmarker.task. Add it to modules/expo-face-landmarker/ios/Resources/."
      )
    }

    let options = FaceLandmarkerOptions()
    options.baseOptions.modelAssetPath = modelPath
    options.runningMode = .image
    options.numFaces = 1
    options.minFaceDetectionConfidence = 0.5
    options.minFacePresenceConfidence = 0.5
    options.minTrackingConfidence = 0.5

    let landmarker = try FaceLandmarker(options: options)
    faceLandmarker = landmarker
    return landmarker
  }

  private static func loadImage(imageUri: String) -> UIImage? {
    if imageUri.hasPrefix("file://"), let url = URL(string: imageUri) {
      return UIImage(contentsOfFile: url.path)
    }

    if FileManager.default.fileExists(atPath: imageUri) {
      return UIImage(contentsOfFile: imageUri)
    }

    return nil
  }
}

File: beautylens/modules/expo-face-landmarker/android/src/main/java/expo/modules/facelandmarker/ExpoFaceLandmarkerModule.kt
package expo.modules.facelandmarker

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.InputStream

class ExpoFaceLandmarkerModule : Module() {
  private var faceLandmarker: FaceLandmarker? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoFaceLandmarker")

    AsyncFunction("detectFromImageAsync") { imageUri: String ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("React context is unavailable")

      val bitmap = loadBitmap(imageUri)
        ?: throw IllegalArgumentException("Could not decode image: $imageUri")

      val landmarker = faceLandmarker ?: createFaceLandmarker().also {
        faceLandmarker = it
      }

      val mpImage = BitmapImageBuilder(bitmap).build()
      val result = landmarker.detect(mpImage)
      val faceLandmarks = result.faceLandmarks().firstOrNull()

      if (faceLandmarks == null) {
        return@AsyncFunction mapOf(
          "status" to "success",
          "face_detected" to false,
          "landmarks" to emptyList<Map<String, Double>>(),
          "num_landmarks" to 0,
          "image_dimensions" to mapOf(
            "width" to bitmap.width,
            "height" to bitmap.height
          ),
          "message" to "No face detected in image"
        )
      }

      val landmarks = faceLandmarks.map { landmark ->
        mapOf(
          "x" to (landmark.x() * bitmap.width).toDouble(),
          "y" to (landmark.y() * bitmap.height).toDouble(),
          "z" to (landmark.z() * bitmap.width).toDouble()
        )
      }

      val xs = landmarks.map { it["x"] ?: 0.0 }
      val ys = landmarks.map { it["y"] ?: 0.0 }
      val minX = xs.minOrNull() ?: 0.0
      val maxX = xs.maxOrNull() ?: 0.0
      val minY = ys.minOrNull() ?: 0.0
      val maxY = ys.maxOrNull() ?: 0.0

      mapOf(
        "status" to "success",
        "face_detected" to true,
        "landmarks" to landmarks,
        "num_landmarks" to landmarks.size,
        "bbox" to mapOf(
          "x" to minX,
          "y" to minY,
          "width" to maxX - minX,
          "height" to maxY - minY
        ),
        "image_dimensions" to mapOf(
          "width" to bitmap.width,
          "height" to bitmap.height
        )
      )
    }
  }

  private fun createFaceLandmarker(): FaceLandmarker {
    val context = appContext.reactContext
      ?: throw IllegalStateException("React context is unavailable")

    val modelName = "face_landmarker.task"
    val assetExists = try {
      context.assets.open(modelName).close()
      true
    } catch (_: Exception) {
      false
    }

    if (!assetExists) {
      throw IllegalStateException(
        "Missing $modelName. Add it to modules/expo-face-landmarker/android/src/main/assets/."
      )
    }

    val baseOptions = BaseOptions.builder()
      .setModelAssetPath(modelName)
      .build()

    val options = FaceLandmarker.FaceLandmarkerOptions.builder()
      .setBaseOptions(baseOptions)
      .setRunningMode(RunningMode.IMAGE)
      .setNumFaces(1)
      .setMinFaceDetectionConfidence(0.5f)
      .setMinFacePresenceConfidence(0.5f)
      .setMinTrackingConfidence(0.5f)
      .build()

    return FaceLandmarker.createFromOptions(context, options)
  }

  private fun loadBitmap(imageUri: String): Bitmap? {
    val context = appContext.reactContext ?: return null
    val uri = Uri.parse(imageUri)

    if (uri.scheme == null) {
      return BitmapFactory.decodeFile(imageUri)
    }

    if (uri.scheme == "file") {
      return BitmapFactory.decodeFile(File(uri.path ?: return null).absolutePath)
    }

    val stream: InputStream = context.contentResolver.openInputStream(uri) ?: return null
    return stream.use { BitmapFactory.decodeStream(it) }
  }
}
